Add Magento Quality Patch package documentation (#7734)
magento/devdocs: src/guides/v2.3/comp-mgr/patching.md

When the [patch][] program is run, this file is read in and the specified changes are made to the file(s).

> There are three types of patches:

Reviewer: If you say three, should it be a numbered list?

Author: I think that's subjective. I prefer using ordered lists when sequence is important.

## Applying patches

> There are three ways to apply patches:

Reviewer: If you say three, should the list be numbered? I don't mind either, just wondering what the threshold is.

Author: I think that's subjective. I prefer using ordered lists when sequence is important.

> There are three ways to apply patches:

Reviewer 2: Does it matter that there are three.  Consider:  You can use any of the following methods to apply patches:

Author: I suppose not. I was just going with what was already there.

-  Using the Magento Quality Patch (MQP) package
-  Using the command line

> -  Using Composer

Author: I actually think it would be better to remove the verb in all three bullets

║ MC-3           │ Fixes issue 3                                   │ Optional │ Not applied │ Required patches:               ║
║                │                                                 │          │             │  - MC-2                         ║
║                │                                                 │          │             │ Affected components:            ║
║                │                                                 │          │             │  - magento/module-cms           ║
╟────────────────┼─────────────────────────────────────────────────┼──────────┼─────────────┼─────────────────────────────────╢
║ MC-3-V2        │ Updated fix for issue 3, replaces MC-3 patch    │ Optional │ N/A         │ Affected components:            ║
║                │                                                 │          │             │  - magento/module-cms           ║
╚════════════════╧═════════════════════════════════════════════════╧══════════╧═════════════╧═════════════════════════════════╝
Magento 2 Enterprise Edition, version 2.3.5.0
```

The status table contains the following types of information:

-  **Type**:

>    -  Optional—All individual patches from the MQP package and the [Magento Cloud Patches]({{ site.baseurl }}/cloud/project/project-patch.html) package are optional for {{ site.data.var.ee }} and {{ site.data.var.ce }} installations.

Reviewer: Same suggestion here, to do something so these stand out a little more?
```suggestion
   -  `Optional`—All individual patches from the MQP package and the [Magento Cloud Patches]({{ site.baseurl }}/cloud/project/project-patch.html) package are optional for {{ site.data.var.ee }} and {{ site.data.var.ce }} installations.
```

>    -  `Deprecate`d—The individual patch is marked as deprecated by Magento and we recommend reverting it if you have applied it. After you revert a deprecated patch, it will no longer be displayed in the status table.

Reviewer 2: ```suggestion
   -  `Deprecated`—Magento has deprecated the individual patch. If you have applied the patch, we recommend that you revert it. The revert operation also removes the patch from the status table.
```

-  **Status**:

   -  `Applied`—The patch has been applied.
   -  `Not applied`—The patch has not been applied.
   -  `N/A`—The status of the patch cannot be defined due to conflicts.

-  **Details**:

   -  `Affected components`—The list of affected Magento modules.

>    -  `Required patches`—The list of required patches (dependencies).

Reviewer 2: What makes a patch required?  

All patches delivered by Cloud Patches package are required.
All patches delivered by MQP are "individual patches" as defined 
Does MQP deliver required patches using Cloud Patches set? 

Maybe `required` isn't a category for Magento open source and enterprise?

Reviewer 3: Required patches are patches that must be applied for the current patch to work properly. It is a dependency. Defined by the published in the MQP configuration file.
https://wiki.corp.magento.com/display/SUP/Advanced+Patch+Configuration+Reference

Author: Clarified in 30a076c7db8f13dcfe631e90168b586ab392f60c

   -  Recommended replacement—The patch that is a recommended replacement for a deprecated patch.

#### Apply individual patches

{:.bs-callout-warning}
We strongly recommend testing all patches in a staging or development environment before deploying to production. We also strongly recommend backing up your data before applying a patch. See [Back up and roll back the file system][].

To apply a single patch, run the following command where `MAGETWO-XXXX` is the patch ID specified in the status table:

```bash
./vendor/bin/magento-patches apply MAGETWO-XXXX
```

> You can also apply several patches at the same time by separating multiple patch IDs with spaces:

Reviewer: ```suggestion
Also, you can apply several patches at the same time by separating each additional patch ID with a space:
```

I make this suggestion for a couple of reasons. So you don't use IDs. And so it is interpreted as "one space"; so not loosely interpreted? Your call.

```bash
./vendor/bin/magento-patches apply MAGETWO-XXXX MAGETWO-YYYY
```

You must clean the cache after applying patches to see changes in the Magento application:

```bash
./bin/magento cache:clean
```

#### Revert individual patches

{:.bs-callout-warning}
We strongly recommend testing all patches in a staging or development environment before deploying to production. We also strongly recommend backing up your data before applying a patch. See [Back up and roll back the file system][].

To revert a single patch, run the following command where `MAGETWO-XXXX` is the patch ID specified in the status table:

```bash
./vendor/bin/magento-patches revert MAGETWO-XXXX
```

> You can also revert several patches at the same time by separating multiple patch IDs with spaces:

Reviewer: ```suggestion
Also, you can revert several patches at the same time by separating each additional patch ID with a space:
```

## Applying patches

You can apply patches using any of the following methods:

-  Magento Quality Patch (MQP) package
-  Command line
-  Composer

{:.bs-callout-info}
To apply a patch to a {{site.data.var.ece}} project, see [Apply patches][].

### Magento Quality Patch (MQP) package {#mqp}

The [MQP package][] delivers individual patches developed by Magento and allows you to apply, revert, and view general information about all individual patches that are available for the installed version of {{ site.data.var.ee }} or {{ site.data.var.ce }}.

{:.bs-callout-warning}
We do not recommend using the MQP package to apply large numbers of patches because it increases the complexity of your code, which makes upgrading to a new version of Magento more difficult.

> 

Reviewer 4: #### Installation prerequisites

MQP requires installed [Git](https://git-scm.com/) to perform operations with patches. See [instructions](https://github.com/git-guides/install-git) on how to install [Git](https://git-scm.com/) on your instance.

## Applying patches

You can apply patches using any of the following methods:

-  Magento Quality Patch (MQP) package
-  Command line
-  Composer

{:.bs-callout-info}
To apply a patch to a {{site.data.var.ece}} project, see [Apply patches][].

### Magento Quality Patch (MQP) package {#mqp}

The [MQP package][] delivers individual patches developed by Magento and allows you to apply, revert, and view general information about all individual patches that are available for the installed version of {{ site.data.var.ee }} or {{ site.data.var.ce }}.

#### Install the MQP package

Add the `magento/quality-patches` Composer package to your `composer.json` file:

```bash
composer require magento/quality-patches
```

```bash
composer update

> ```

Author: Done

## Applying patches

You can apply patches using any of the following methods:

-  Magento Quality Patch (MQP) package
-  Command line
-  Composer

{:.bs-callout-info}
To apply a patch to a {{site.data.var.ece}} project, see [Apply patches][].

### Magento Quality Patch (MQP) package {#mqp}

The [MQP package][] delivers individual patches developed by Magento and allows you to apply, revert, and view general information about all individual patches that are available for the installed version of {{ site.data.var.ee }} or {{ site.data.var.ce }}.

{:.bs-callout-warning}
We do not recommend using the MQP package to apply large numbers of patches because it increases the complexity of your code, which makes upgrading to a new version of Magento more difficult.

#### Install the MQP package

{:.bs-callout-info}

> You must install [Git](https://github.com/git-guides/install-git) before installing the MQP package.

Reviewer 4: Usually, Git is installed on servers. So maybe it will be good to add "You must install [Git](https://github.com/git-guides/install-git) if it hasn't installed yet"

Author: Done 136e4a089ca9cc23d435e9cc74c70474f1f34d7f
